### PR TITLE
feat: subgraph integration

### PR DESCRIPTION
**Pull request by jmulq**
This goes alongside the following updates to the subgraph:
- https://github.com/Vault-Hill/vhc-marketplace-subgraph/pull/5

### CODE REVIEW

File: src/services/subgraph.service/subgraph.service.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        query: gql(collectionsQuery)
+      })
+      const collections: Collection[] = []
+      for (const col of data.collections) {
+        const collection = await this.buildCollection(col)
+        if (collection) collections.push(collection)
+      }
+      return collections
+    } catch (error) {
+      // TODO: Add error logging
```
Comment (anjorinjnr) [inline src/services/subgraph.service/subgraph.service.ts:36]: this todo is redundant, when you console.log[error] you are indeed logging. we can replace console.log with something more sophisticated to log object with properties etc but this is already logging stuff and it's fine for now.

File: src/services/subgraph.service/types.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import { CollectionType } from '../../graphql-types/enums'
+
+export type CollectionVland = {
```
Comment (anjorinjnr) [inline src/services/subgraph.service/types.ts:3]: we should discuss this types. i don't understand the difference between collectionvland and subgraphvland, perhaps you can add some comments to the types

File: src/services/subgraph.service/query.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        owner {
+          id
+        }
+      }
+      createdAtTimestamp
+    }
+  }
+`
+
+export const collectionQuery = `
```
Comment (anjorinjnr) [inline src/services/subgraph.service/query.ts:41]: collectionByAddressQuery?

File: src/application.ts [hunk 4]
```diff
@@ -24,6 +27,9 @@ export class VhcMarketplaceServerApplication extends BootMixin(
     // Validate env variables
     this.validateEnvironment()
 
+    // Validate env variables
```
Comment (anjorinjnr) [inline src/application.ts:30]: why is this repeated?

File: package.json [hunk 5]
```diff
@@ -57,12 +57,14 @@
     "@loopback/rest": "^11.1.1",
     "@loopback/rest-explorer": "^4.1.1",
     "@loopback/service-proxy": "^4.1.1",
+    "@types/faker": "^6.6.9",
```
Comment (anjorinjnr) [inline package.json:60]: we can remove this.

File: src/repositories/asset.repository.ts [hunk 6]
```diff
@@ -1,20 +1,18 @@
...
 
-export class AssetRepository extends DefaultCrudRepository<
-  Asset,
-  typeof Asset.prototype.id,
-  {}
-> {
+@lifeCycleObserver('repository')
+export class AssetRepository
+  extends DefaultCrudRepository<Asset, typeof Asset.prototype.id, {}>
+  implements LifeCycleObserver
```
Comment (anjorinjnr) [inline src/repositories/asset.repository.ts:9]: remove the implements

File: src/repositories/asset.repository.ts [hunk 7]
```diff
@@ -1,20 +1,18 @@
...
   constructor(@inject('datasources.inmemory') dataSource: InmemoryDataSource) {
     super(Asset, dataSource)
   }
 
-  async add(data: Partial<Asset>) {
-    const asset = new Asset(data)
-    asset.id = Date.now().toString()
-    return this.create(data)
-  }
+  start() {}
```
Comment (anjorinjnr) [inline src/repositories/asset.repository.ts:15]: remove both methods

File: src/services/subgraph.service/subgraph.service.ts [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    @inject(AppBindings.APOLLO_CLIENT)
+    private client: ApolloClient<NormalizedCacheObject>
+  ) {}
+
+  async getCollections() {
+    try {
+      const { data } = await this.client.query({
+        query: gql(collectionsQuery)
+      })
+      const collections: Collection[] = []
```
Comment (anjorinjnr) [inline src/services/subgraph.service/subgraph.service.ts:29]: the buildCollection is specifically about vLand, so unless there can be multiple vland collections am not sure why we have this for loop.

File: src/services/subgraph.service/subgraph.service.ts [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    } catch (error) {
+      // TODO: Add error logging
+      console.error(error)
+    }
+  }
+
+  private async buildVlandAsset(vlandToken: SubgraphVland) {
+    try {
+      const uri = vlandToken.metadataURI
+      const response = await fetch(uri)
```
Comment (anjorinjnr) [inline src/services/subgraph.service/subgraph.service.ts:103]: we are gonna run into latency issues if every request involves

app -> server -> graph       -> app
                         -> ipfs fetch

good to idea to open an issue to revisit this later

File: src/application.ts [hunk 10]
```diff
@@ -8,13 +8,16 @@ import {
...
 import { MySequence } from './sequence'
+import { AppBindings } from './app.bindings'
+import { ApolloClient, HttpLink, InMemoryCache } from '@apollo/client/core'
+import { isEmpty } from 'lodash'
+import fetch from 'cross-fetch'
 
 export { ApplicationConfig }
 
-const ENV_VARS: string[] = []
+const ENV_VARS = ['SUBGRAPH_API_URL']
```
Comment (anjorinjnr) [inline src/application.ts:20]: do/will we have dev instance of the subgraph vs prod.  please provide url of the instance you're using for test.

**Review by anjorinjnr — CHANGES_REQUESTED** [review]
This enough code that certain components should have unit tests, lets discuss that if you need help.

File: src/repositories/asset.repository.ts (hunk 7 above)
Comment (jmulq) [reply src/repositories/asset.repository.ts:15]: Removed from asset and collection repo

File: src/repositories/asset.repository.ts (hunk 6 above)
Comment (jmulq) [reply src/repositories/asset.repository.ts:9]: Removed implements. Presume it's also good to remove the lifecycle observer decorator?

File: src/application.ts (hunk 10 above)
Comment (jmulq) [reply src/application.ts:20]: https://api.thegraph.com/subgraphs/name/vault-hill/vhc-marketplace dev version.

we should have one for dev, one for staging, and one for prod I would say.

File: src/services/subgraph.service/subgraph.service.ts (hunk 1 above)
Comment (jmulq) [reply src/services/subgraph.service/subgraph.service.ts]: Remove in latest PR. Logger ticket for future enhancement -> https://github.com/Vault-Hill/vhc-marketplace-server/issues/18

File: src/services/subgraph.service/types.ts (hunk 2 above)
Comment (jmulq) [reply src/services/subgraph.service/types.ts]: Have made the types for explanatory. They were worded badly beforehand.

The subgraph types are the `Collection` and `Vland` types coming from the subgraph.

The metadata type is the vland metadata stored on IPFS.

File: src/services/subgraph.service/query.ts (hunk 3 above)
Comment (jmulq) [reply src/services/subgraph.service/query.ts]: Good shout - changed this.

File: src/application.ts (hunk 4 above)
Comment (jmulq) [reply src/application.ts]: Just to be doubly sure we validate the env.

Comment (jmulq) [reply src/application.ts]: jk, removed in latest push

File: src/services/subgraph.service/subgraph.service.ts (hunk 8 above)
Comment (jmulq) [reply src/services/subgraph.service/subgraph.service.ts]: Good point. Changed it in latest push. Still returns an array, just without the loop. Can update when more collections come in.

File: src/services/subgraph.service/subgraph.service.ts (hunk 9 above)
Comment (jmulq) [reply src/services/subgraph.service/subgraph.service.ts:95]: Raised to eng tracker and added issue here https://github.com/Vault-Hill/vhc-marketplace-server/issues/19

**Review by anjorinjnr — DISMISSED** [review]

**Review by abbeydabiri — APPROVED** [review]

**Pull request merged**